unit/mr_test.c

```diff
@@ -48,105 +48,158 @@ static char err_buf[512];
+				goto fail;
+			}
 
-		// TODO test for various access
-		ret = fi_mr_reg(domain, buf, buf_size, ft_info_to_mr_access(fi), 0,
-				FT_MR_KEY, 0, &mr, NULL);
-		if (ret) {
-			FT_UNIT_STRERR(err_buf, "fi_mr_reg failed", ret);
-			goto fail;
-		}
-		ret = fi_close(&mr->fid);
-		if (ret) {
-			FT_UNIT_STRERR(err_buf, "fi_close failed", ret);
-			goto fail;
+			ret = fi_close(&mr->fid);
+			if (ret) {
+				FT_UNIT_STRERR(err_buf, "fi_close failed", ret);
+				goto fail;
+			}
 		}
 	}
 	testret = PASS;
 fail:
-	mr = NULL;
+	ft_free_bit_combo(access_combinations);
 	return TEST_RET_VAL(ret, testret);
 }
 
 static int mr_regv()
 {
-	int ret;
-	int testret;
+	int i, j, n;
+	int ret = 0;
+	int testret = FAIL;
 	struct fid_mr *mr;
-	struct iovec iov;
+	struct iovec *iov;
+	char *base;
 
-	testret = FAIL;
-
-	iov.iov_base = buf;
-	iov.iov_len = test_size[test_cnt / 2].size;
-
-	// TODO test for various iov counts
-	ret = fi_mr_regv(domain, &iov, 1, ft_info_to_mr_access(fi), 0,
-			FT_MR_KEY, 0, &mr, NULL);
-	if (ret) {
-		FT_UNIT_STRERR(err_buf, "fi_mr_regv failed", ret);
+	if (!fi->domain_attr->mr_iov_limit) {
+		ret = 1;
+		FT_UNIT_STRERR(err_buf, "fi_mr_regv failed, mr_iov_limit", ret);
 		goto fail;
 	}
-	ret = fi_close(&mr->fid);
-	if (ret) {
-		FT_UNIT_STRERR(err_buf, "fi_close failed", ret);
-		goto fail;
+
+	iov = calloc(fi->domain_attr->mr_iov_limit, sizeof(struct iovec));
+	if (!iov) {
+		perror("calloc");
+		return FI_ENOMEM;
+	}
+
+	for (i = 0; i < test_cnt &&
+	     test_size[i].size <= fi->domain_attr->mr_iov_limit; i++) {
+		n = test_size[i].size;
+		base = buf;
+
+		for (j = 0; j < n; j++) {
+			iov[j].iov_base = base;
+			iov[j].iov_len = test_size[test_cnt - 1].size / n;
+			base += iov[j].iov_len;			
+		}
+
+		ret = fi_mr_regv(domain, &iov[0], n, ft_info_to_mr_access(fi), 0,
+				 FT_MR_KEY, 0, &mr, NULL);
+		if (ret) {
+			FT_UNIT_STRERR(err_buf, "fi_mr_regv failed", ret);
+			goto clean;
+		}
+
+		ret = fi_close(&mr->fid);
+		if (ret) {
+			FT_UNIT_STRERR(err_buf, "fi_close failed", ret);
+			goto clean;
+		}
 	}
 	testret = PASS;
+clean:
+	free(iov);
 fail:
-	mr = NULL;
 	return TEST_RET_VAL(ret, testret);
 }
 
 static int mr_regattr()
 {
-	int ret;
-	int testret;
+	int i, j, n;
+	int ret = 0;
+	int testret = FAIL;
 	struct fid_mr *mr;
-	struct iovec iov;
+	struct iovec *iov;
 	struct fi_mr_attr attr;
+	char *base;
 
-	testret = FAIL;
-
-	iov.iov_base = buf;
-	iov.iov_len = test_size[test_cnt / 2].size;
+	if (!fi->domain_attr->mr_iov_limit) {
+		ret = 1;
+		FT_UNIT_STRERR(err_buf, "fi_mr_regattr failed, mr_iov_limit", ret);
+		goto fail;
+	}
 
-	attr.mr_iov = &iov;
-	attr.iov_count = 1;
 	attr.access = ft_info_to_mr_access(fi);
 	attr.requested_key = FT_MR_KEY;
 	attr.context = NULL;
-
-	// TODO test for various iov sizes
-	ret = fi_mr_regattr(domain, &attr, 0, &mr);
-	if (ret) {
-		FT_UNIT_STRERR(err_buf, "fi_mr_regattr failed", ret);
-		goto fail;
+	
+	iov = calloc(fi->domain_attr->mr_iov_limit, sizeof(struct iovec));
```

Comment:
nit: In most calloc/malloc calls it's preferable to use sizeof(var) or sizeof(*var) instead of sizeof(struct blah).  It makes it easier to maintain.  This comment applies to all calloc calls in this series.